### PR TITLE
Reset password forms

### PR DESCRIPTION
**Pull request by aydevelop**

### CODE REVIEW

File: backend/Recodo.API/Controllers/AuthController.cs [hunk 1]
```diff
@@ -62,5 +62,6 @@ public async Task<IActionResult> GoogleLogin([FromBody] ExternalAuthDto external
 
             return new JsonResult(result);
         }
+
```
Comment (Anatolii-Ivanov) [inline backend/Recodo.API/Controllers/AuthController.cs:65]: Not needed

File: backend/Recodo.API/Controllers/UserController.cs [hunk 2]
```diff
@@ -1,22 +1,39 @@
...
+        private readonly UserService _userService;
+
+        public UserController(AuthService authService, UserService userService)
         {
+            _authService = authService;
             _userService = userService;
         }
 
+        [HttpPost("ResetPassword")]
+        public async Task<IActionResult> ResetPassword(UpdateUserDTO user)
```
Comment (Anatolii-Ivanov) [inline backend/Recodo.API/Controllers/UserController.cs]: Why you send updateUserDTO model? It should be only email for this request

File: backend/Recodo.API/Controllers/UserController.cs [hunk 3]
```diff
@@ -1,22 +1,39 @@
...
+        {
+            await _userService.ResetPassword(user.Email);
+            return NoContent();
+        }
+
+        [HttpPost("ResetPasswordDone")]
+        public async Task<IActionResult> ResetPasswordDone(UpdateUserDTO user)
```
Comment (Anatolii-Ivanov) [inline backend/Recodo.API/Controllers/UserController.cs]: The same question

File: backend/Recodo.API/Controllers/UserController.cs [hunk 4]
```diff
@@ -1,22 +1,39 @@
...
         }
 
+        [HttpPost("ResetPassword")]
+        public async Task<IActionResult> ResetPassword(UpdateUserDTO user)
+        {
+            await _userService.ResetPassword(user.Email);
+            return NoContent();
+        }
+
+        [HttpPost("ResetPasswordDone")]
```
Comment (Anatolii-Ivanov) [inline backend/Recodo.API/Controllers/UserController.cs]: Rename to process (for now it's a finished action). The same for methods. They should show what they do, not in what they will result

File: backend/Recodo.BLL/Services/EmailService.cs [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+            using (var client = new SmtpClient())
+            {
+                await client.ConnectAsync(configuration["SmtpServer"], Convert.ToInt32(configuration["SmtpPort"]), false);
+                await client.AuthenticateAsync(configuration["SmtpLogin"], configuration["SmtpPassword"]);
+                await client.SendAsync(emailMessage);
+                await client.DisconnectAsync(true);
+            }
+
+            Console.WriteLine(message);
```
Comment (Anatolii-Ivanov) [inline backend/Recodo.BLL/Services/EmailService.cs]: I think you know

File: backend/Recodo.BLL/Services/EmailService.cs [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+            emailMessage.From.Add(new MailboxAddress("admin@recodo.com", "admin@recodo.com"));
+            emailMessage.To.Add(new MailboxAddress(email, email));
+            emailMessage.Subject = subject;
+            emailMessage.Body = new TextPart(MimeKit.Text.TextFormat.Text)
+            {
+                Text = message
+            };
+
+            using (var client = new SmtpClient())
```
Comment (Anatolii-Ivanov) [inline backend/Recodo.BLL/Services/EmailService.cs]: I'm not really happy about using SmtpClient, users will often get such email in spam...

File: backend/Recodo.BLL/Services/UserService.cs [hunk 7]
```diff
@@ -14,15 +16,22 @@
...
 using System.Threading.Tasks;
+using Thread_.NET.BLL.Services;
 
 namespace Recodo.BLL.Services
 {
     public class UserService : BaseService
     {
+        private readonly AuthService _authService;
+        private readonly UserService _userService;
+        private readonly IConfiguration _configuration;
```
Comment (Anatolii-Ivanov) [inline backend/Recodo.BLL/Services/UserService.cs:23]: Some github actions

File: backend/Recodo.Common/Dtos/User/UpdateUserDTO.cs [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+    public class UpdateUserDTO
+    {
+        public int Id { get; set; }
+        public string Email { get; set; }
+        public string WorkspaceName { get; set; }
+        public string AvatarLink { get; set; } = "";
+        public string PasswordNew { get; set; }
+        public string PasswordCurrent { get; set; }
+        public string Token { get; set; }
+    }
```
Comment (Anatolii-Ivanov) [inline backend/Recodo.Common/Dtos/User/UpdateUserDTO.cs]: Why you need it?

File: frontend/src/app/modules/auth/register-form/register-form.component.ts [hunk 9]
```diff
@@ -32,37 +33,41 @@ export class RegisterFormComponent {
...
-      workspaceName: [,
-          [Validators.required,
-          Validators.pattern('^[а-яА-ЯёЁa-zA-Z\`\'][а-яА-ЯёЁa-zA-Z-\`\' ]+[а-яА-ЯёЁa-zA-Z\`\']?$'),
-          Validators.minLength(3),
-          Validators.maxLength(30),
-          startsOrEndWithSpace
+    this.registerForm = this.formBuilder.group(
+      {
+        workspaceName: [
+          ,
```
Comment (Anatolii-Ivanov) [inline frontend/src/app/modules/auth/register-form/register-form.component.ts]: Can you remove this comma?

File: frontend/src/app/modules/auth/register-form/register-form.component.ts [hunk 10]
```diff
@@ -32,37 +33,41 @@ export class RegisterFormComponent {
...
-          Validators.required,
-          Validators.minLength(8),
-          Validators.maxLength(20),
-          Validators.pattern(
-            '^(?=.*[0-9])(?=.*[a-zA-Z])([a-zA-Z0-9*.!@\\#$%^&`\\(\\)\\{\\}\\[\\]\\\\:;<>,‘.?/~_+-=|]+)$'
-            ),
-          cannotContainSpace
+        confirmPassword: [, [Validators.required]],
+        password: [
+          ,
```
Comment (Anatolii-Ivanov) [inline frontend/src/app/modules/auth/register-form/register-form.component.ts]: All of them

File: frontend/src/app/modules/auth/reset-done-page/reset-done-page.component.ts [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+  public currentUser: UserDto = {} as UserDto;
+
+  constructor(
+    private route: ActivatedRoute,
+    private router: Router,
+    private formBuilder: FormBuilder,
+    private registrationService: RegistrationService,
+    private externalAuthService: ExternalAuthService,
+    private userService: UserService
+  ) {
```
Comment (Anatolii-Ivanov) [inline frontend/src/app/modules/auth/reset-done-page/reset-done-page.component.ts]: You don't need all of them

File: frontend/src/app/modules/auth/reset-pass-page/reset-pass-page.component.ts [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+  public hideConfirmPass = true;
+  public currentUser: UserDto = {} as UserDto;
+  public isDone: boolean = false;
+
+  constructor(
+    private router: Router,
+    private formBuilder: FormBuilder,
+    private loginService: LoginService,
+    private externalAuthService: ExternalAuthService,
+    private userService: UserService
```
Comment (Anatolii-Ivanov) [inline frontend/src/app/modules/auth/reset-pass-page/reset-pass-page.component.ts]: the same

File: frontend/src/app/services/user.service.ts [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+    return '/User';
+  }
+
+  public resetPassword(email: string) {
+    return this.addWithUrl<{ email: string }, {}>('Resetpassword/', {
+      email,
+    });
+  }
+
+  public resetPasswordDone(email: string, password: string) {
```
Comment (Anatolii-Ivanov) [inline frontend/src/app/services/user.service.ts]: Rename function

File: backend/Recodo.BLL/Services/UserService.cs [hunk 14]
```diff
@@ -76,5 +85,37 @@ public async Task<UserDTO> GetUserById(int userId)
...
+            {
+                throw new ExistUserException(user.Email);
+            }
+
+            var salt = SecurityHelper.GetRandomBytes();
+            existUser.Salt = Convert.ToBase64String(salt);
+            existUser.Password = SecurityHelper.HashPassword(user.PasswordNew, salt);
+
+            _context.Users.Update(existUser);
+            await _context.SaveChangesAsync();
```
Comment (Anatolii-Ivanov) [inline backend/Recodo.BLL/Services/UserService.cs:177]: Sign in user to system after this process

File: backend/Recodo.API/Controllers/AuthController.cs (hunk 1 above)
Comment (aydevelop) [reply backend/Recodo.API/Controllers/AuthController.cs:65]: Deleted

File: backend/Recodo.API/Controllers/UserController.cs (hunk 2 above)
Comment (aydevelop) [reply backend/Recodo.API/Controllers/UserController.cs]: Fixed

File: backend/Recodo.API/Controllers/UserController.cs (hunk 3 above)
Comment (aydevelop) [reply backend/Recodo.API/Controllers/UserController.cs]: Fixed

File: backend/Recodo.API/Controllers/UserController.cs (hunk 4 above)
Comment (aydevelop) [reply backend/Recodo.API/Controllers/UserController.cs]: Renamed

File: backend/Recodo.BLL/Services/EmailService.cs (hunk 5 above)
Comment (aydevelop) [reply backend/Recodo.BLL/Services/EmailService.cs]: Just logger, deleted

File: backend/Recodo.BLL/Services/EmailService.cs (hunk 6 above)
Comment (aydevelop) [reply backend/Recodo.BLL/Services/EmailService.cs]: Migrated to SendGrid

File: backend/Recodo.BLL/Services/UserService.cs (hunk 7 above)
Comment (aydevelop) [reply backend/Recodo.BLL/Services/UserService.cs:23]: Deleted

File: backend/Recodo.Common/Dtos/User/UpdateUserDTO.cs (hunk 8 above)
Comment (aydevelop) [reply backend/Recodo.Common/Dtos/User/UpdateUserDTO.cs]: Deleted

File: frontend/src/app/modules/auth/register-form/register-form.component.ts (hunk 9 above)
Comment (aydevelop) [reply frontend/src/app/modules/auth/register-form/register-form.component.ts]: Prettier works

File: frontend/src/app/modules/auth/register-form/register-form.component.ts (hunk 10 above)
Comment (aydevelop) [reply frontend/src/app/modules/auth/register-form/register-form.component.ts]: Prettier works

File: frontend/src/app/modules/auth/reset-done-page/reset-done-page.component.ts (hunk 11 above)
Comment (aydevelop) [reply frontend/src/app/modules/auth/reset-done-page/reset-done-page.component.ts]: Deleted

File: frontend/src/app/modules/auth/reset-pass-page/reset-pass-page.component.ts (hunk 12 above)
Comment (aydevelop) [reply frontend/src/app/modules/auth/reset-pass-page/reset-pass-page.component.ts]: Deleted

File: frontend/src/app/services/user.service.ts (hunk 13 above)
Comment (aydevelop) [reply frontend/src/app/services/user.service.ts]: Ok

File: backend/Recodo.BLL/Services/UserService.cs (hunk 14 above)
Comment (aydevelop) [reply backend/Recodo.BLL/Services/UserService.cs:177]: Done

**Comment by aydevelop** [comment]
Please, use binary credentials in appsettings.json -  "SendGridKey": "",   "SendGridFromEmail": ""

File: backend/Recodo.BLL/Services/AuthService.cs [hunk 15]
```diff
@@ -38,7 +38,7 @@ public async Task<AuthUserDTO> Authorize(LoginUserDTO userDTO)
                 throw new NotFoundException(nameof(User));
             }
 
-            if (!SecurityHelper.IsValidPassword(userEntity.Password, userDTO.Password, userEntity.Salt))
+            if (checkPass == true && !SecurityHelper.IsValidPassword(userEntity.Password, userDTO.Password, userEntity.Salt))
```
Comment (Anatolii-Ivanov) [inline backend/Recodo.BLL/Services/AuthService.cs:41]: Seriously?
 if (property == true)???

File: frontend/src/app/modules/auth/reset-done-page/reset-done-page.component.ts [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+      this.email = decoded.email;
+    });
+  }
+
+  private validateForm() {
+    this.registerForm = this.formBuilder.group(
+      {
+        confirmPassword: [, [Validators.required]],
+        password: [
+          ,
```
Comment (Anatolii-Ivanov) [inline frontend/src/app/modules/auth/reset-done-page/reset-done-page.component.ts:52]: Commas?

**Review by Anatolii-Ivanov — CHANGES_REQUESTED** [review]

**Comment by aydevelop** [comment]
Fixed

File: backend/Recodo.BLL/Services/AuthService.cs (hunk 15 above)
Comment (Anatolii-Ivanov) [reply backend/Recodo.BLL/Services/AuthService.cs:41]: Again

**Review by Anatolii-Ivanov — APPROVED** [review]

**Pull request merged**